edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/ScopeIdentitiesHandler.cs

```diff
@@ -0,0 +1,134 @@
+    using System.Linq;
+    using System.Text;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Devices.Edge.Hub.Core;
+    using Microsoft.Azure.Devices.Edge.Hub.Core.Identity.Service;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Azure.Devices.Edge.Util.Concurrency;
+    using Microsoft.Extensions.Logging;
+    using Newtonsoft.Json;
+
+    public class ScopeIdentitiesHandler : IMessageProducer
+    {
+        const string AddOrUpdateTopic = "$internal/identities/addOrUpdate";
+        const string RemoveTopic = "$internal/identities/removeTopic";
+        readonly IDeviceScopeIdentitiesCache deviceScopeIdentitiesCache;
+        readonly AtomicBoolean connected = new AtomicBoolean(false);
+        IMqttBrokerConnector connector;
+        Dictionary<string, BrokerServiceIdentity> brokerServiceIdentities = new Dictionary<string, BrokerServiceIdentity>();
+
+        public ScopeIdentitiesHandler(IDeviceScopeIdentitiesCache deviceScopeIdentitiesCache)
+        {
+            this.deviceScopeIdentitiesCache = deviceScopeIdentitiesCache;
+            this.deviceScopeIdentitiesCache.ServiceIdentityUpdated += (sender, serviceIdentity) => Task.WhenAll(this.ServiceIdentityUpdated(serviceIdentity));
+            this.deviceScopeIdentitiesCache.ServiceIdentityRemoved += (sender, identity) => Task.WhenAll(this.ServiceIdentityRemoved(identity));
+        }
+
+        public void SetConnector(IMqttBrokerConnector connector)
+        {
+            this.connector = connector;
+            this.connector.OnConnected += (sender, args) => Task.WhenAll(this.OnConnect());
+        }
+
+        async Task OnConnect()
+        {
+            this.connected.Set(true);
+            await this.PublishAddOrUpdateBrokerServiceIdentities(this.brokerServiceIdentities);
+        }
+
+        async Task ServiceIdentityRemoved(string identity)
+        {
+            if (this.connected.Get())
+            {
+                await this.PublishRemoveBrokerIdentityService(identity);
+            }
+            else
+            {
+                this.brokerServiceIdentities.Remove(identity);
+            }
+        }
+
+        async Task ServiceIdentityUpdated(ServiceIdentity serviceIdentity)
+        {
+            BrokerServiceIdentity brokerServiceIdentity = new BrokerServiceIdentity(serviceIdentity.Id, await this.deviceScopeIdentitiesCache.GetAuthChain(serviceIdentity.Id));
+            if (this.connected.Get())
+            {
+                await this.PublishAddOrUpdateBrokerServiceIdentity(brokerServiceIdentity);
+            }
+            else
+            {
+                this.brokerServiceIdentities.Add(brokerServiceIdentity.Identity, brokerServiceIdentity);
+            }
+        }
+
+        async Task PublishAddOrUpdateBrokerServiceIdentities(Dictionary<string, BrokerServiceIdentity> brokerServiceIdentities)
+        {
+            foreach (BrokerServiceIdentity brokerServiceIdentity in brokerServiceIdentities.Values)
```

Comment:
can we batch send the identities instead? For individual Add/Update we can send an array of 1 element.